### ISSUE TITLE
Syncing mjData.qpos and mjData.xpos: additional forward kinematics pass required after simulation step

### ISSUE THREAD

**Issue by aldethux** [deepmind/mujoco#498]
Hello, I have some questions about #430:

1. What is the rationale behind this design?
I see that it's easier to make preparations in the same function just before integration, as this allows to call `mj_step` whether `mjData` has been changed from outside or not. However, in my experience, usually what is needed is a consistent state. Most models I work with expose their state straight from `mjData`, which leads to inconsistency in `qpos` / `xpos`.

2. Is calling `mj_forward` after `mj_step` the best way to solve this issue with `mujoco-py` (there are no `mj_step1`, `mj_step2`, `mj_kinematics`)?

4. Is it a good idea to introduce a new function that acts like `mj_step` but cyclically shifts pipeline s.t. integration is always followed by a forward kinematics pass during a single call?

**Comment by yuvaltassa** [comment]
1. The rational is precisely what you are alluding to. `mj_step` takes the minimal computation required to get you to the next state. If you want to compute things about this new state then it is up to you. Note that if you don't read from `qpos` and `xpos` but rather only values from sensors, then everything will be self-consistent, albeit delayed by 1 timestep w.r.t to the new state. The correct way to think about this is that sensor values are associated with a state transition, rather than a state.

2. Obviously we cannot help you with mujoco-py. Sorrry about that.

3. It is a good idea, and it already exists. It is called `mj_step1` and `mj_step2`. If you examine the implementation of `Physics.step` in dm_control, this precisely [what it does](https://github.com/deepmind/dm_control/blob/main/dm_control/mujoco/engine.py#L162). When you set the state with `reset_context` you call `mj_forward`, and then afterwards you take a step by calling `mj_step2` and then `mj_step1`, in "reverse order", effectively shifting the integration pipeline.

While we used this approach for many years, we are now backing away from it and going back to regular `mj_step`. The reason is the following. While keeping position and velocity-dependent sensors (and some `mjData` values) consistent with the new state is possible, it is impossible to do so with acceleration/force dependent sensors (e.g. accelerometer/touch sensors). The latter values are only computed in the step2 part (e.g. are dependent on the controls). So the step2->step1 approach has the downside that some sensors are w.r.t to the new state and some are w.r.t the old state. Imagine having an IMU unit where the gyroscope is always 1 step ahead of the accelerometer. This is the situation in `dm_control`, and we think it is more confusing and ultimately less good.

A similar approach which might be considered is to e.g. flip `mj_forward` and `mj_Euler`. So you would start by just integrating the old accelerations, and then call `mj_forward` to update all the values. This would have the strange effect that current controls no longer affect the current state transition, but only the next one.

To sum up:
1. There is no perfect solution, they all have drawbacks. We prefer the standard `mj_step` due to its conceptual simplicity.
2. MuJoCo's reconfigurable allows you to easily try all these options yourself.
3. Stop using `mujoco-py`.
